internal/leadership/elector.go

```diff
@@ -62,206 +89,294 @@ type Elector struct {
+		if err != nil { //nolint:staticcheck
+			// TODO(brandur): Propagate this after refactor.
+		}
+	}
+
+	go func() {
+		// This defer should come first so that it's last out, thereby avoiding
+		// races.
+		defer close(stopped)
+
+		e.Logger.InfoContext(ctx, e.Name+": Run loop started")
+		defer e.Logger.InfoContext(ctx, e.Name+": Run loop stopped")
+
+		if sub != nil {
+			defer sub.Unlisten(ctx)
+		}
+
+		for {
+			if err := e.attemptGainLeadershipLoop(ctx); err != nil {
+				// Function above only returns an error if context was cancelled
+				// or overall context is done.
+				if !errors.Is(err, context.Canceled) && ctx.Err() == nil {
+					panic(err)
+				}
 				return
-			default:
-				// TODO: backoff
-				e.logger.Error("error keeping leadership", "err", err)
-				continue
+			}
+
+			e.Logger.Info(e.Name+": Gained leadership", "client_id", e.clientID)
+			e.testSignals.GainedLeadership.Signal(struct{}{})
+
+			err := e.keepLeadershipLoop(ctx)
+			if err != nil {
+				if errors.Is(err, context.Canceled) {
+					return
+				}
+
+				if !errors.Is(err, errLostLeadership) {
+					e.Logger.Error(e.Name+": Error keeping leadership", "client_id", e.clientID, "err", err)
+				}
 			}
 		}
-	}
+	}()
+
+	return nil
 }
 
-func (e *Elector) gainLeadership(ctx context.Context, leadershipNotificationChan <-chan struct{}) bool {
+func (e *Elector) attemptGainLeadershipLoop(ctx context.Context) error {
+	var numErrors int
+
 	for {
-		success, err := e.attemptElect(ctx)
-		if err != nil && !errors.Is(err, context.Canceled) {
-			e.logger.Error("error attempting to elect", "err", err)
+		e.Logger.Info(e.Name+": Attempting to gain leadership", "client_id", e.clientID)
+
+		elected, err := attemptElectOrReelect(ctx, e.exec, false, &riverdriver.LeaderElectParams{
+			LeaderID: e.clientID,
+			Name:     e.instanceName,
+			TTL:      e.ttl,
+		})
+		if err != nil {
+			if errors.Is(err, context.Canceled) || ctx.Err() != nil {
+				return err
+			}
+
+			numErrors++
+			e.Logger.Error(e.Name+": Error attempting to elect", "client_id", e.clientID, "err", err, "num_errors", numErrors)
+			e.CancellableSleepExponentialBackoff(ctx, numErrors-1, baseservice.MaxAttemptsBeforeResetDefault)
+			continue
 		}
-		if success {
-			return true
+		if elected {
+			return nil
 		}
 
+		numErrors = 0
+
+		e.Logger.Info(e.Name+": Leadership bid was unsuccessful (not an error)", "client_id", e.clientID)
+		e.testSignals.DeniedLeadership.Signal(struct{}{})
+
 		select {
-		case <-ctx.Done():
-			return false
-		case <-time.After(e.interval):
-			// TODO: This could potentially leak memory / timers if we're seeing a ton
-			// of resignations. May want to make this reusable & cancel it when retrying?
-		case <-leadershipNotificationChan:
-			// Somebody just resigned, try to win the next election immediately.
+		case <-e.CancellableSleepRandomBetweenC(ctx, e.electInterval, e.electInterval+e.electIntervalJitter):
```

Comment:
there's a potential leaking of timers here, but only if resignations are happening frequently

Reply:
Yeah, good point. I just added the comment back that was previously there.

I think it might be kind of neat if we could do a `timeutil` utility exactly like `Ticker` but which would tick within a random range. I'll put it on my TODO list.